File: Artemis2MenuBar/MissionData.swift
import Foundation

struct MissionPhaseInfo: Equatable {
    var id: String = "prelaunch"
    var name: String = "Pre-Launch"
    var description: String = ""
    var icon: String = "clock"
}

struct MissionData {
    var missionElapsedTime: TimeInterval = 0
    var distanceFromEarth: Double = 0       // km altitude (from surface)
    var distanceFromMoon: Double = 0        // km
    var speed: Double = 0                   // km/s
    var phase = MissionPhaseInfo()
    var crewMembers: [String] = []
    var missionName: String = ""
    var missionSubtitle: String = ""

    /// Dynamic metric values keyed by source ID
    var values: [String: MetricValue] = [:]

    /// Compute all built-in metric values from the current state
    mutating func updateBuiltInValues() {
        values["met"] = .duration(missionElapsedTime)
        values["altitude"] = .number(distanceFromEarth)
        values["moonDistance"] = .number(distanceFromMoon)
        values["speed"] = .number(speed)
        values["phase"] = .text(phase.name)
    }

    /// Get the formatted value for a metric definition
    func formattedValue(for metric: Metric, units: UnitSystem) -> String {
        guard let value = values[metric.source] else { return "—" }
        return value.format(as: metric.format, units: units)
    }

    /// Get the detailed formatted value for a metric definition
    func formattedDetailValue(for metric: Metric, units: UnitSystem) -> String {
        guard let value = values[metric.source] else { return "—" }
        return value.formatDetail(as: metric.format, units: units)
    }
}

File: OrbitalWindows/Models/MissionData.cs
namespace Orbital.Models;

public class MissionPhaseInfo
{
    public string Id { get; set; } = "prelaunch";
    public string Name { get; set; } = "Pre-Launch";
    public string Description { get; set; } = "";
    public string Icon { get; set; } = "clock";
}

public class MissionData
{
    public double MissionElapsedTime { get; set; }
    public double DistanceFromEarth { get; set; }  // km altitude (from surface)
    public double DistanceFromMoon { get; set; }    // km
    public double Speed { get; set; }               // km/s
    public MissionPhaseInfo Phase { get; set; } = new();
    public List<string> CrewMembers { get; set; } = [];
    public string MissionName { get; set; } = "";
    public string MissionSubtitle { get; set; } = "";

    public Dictionary<string, MetricValue> Values { get; set; } = new();

    public void UpdateBuiltInValues()
    {
        Values["met"] = new DurationValue(MissionElapsedTime);
        Values["altitude"] = new NumberValue(DistanceFromEarth);
        Values["moonDistance"] = new NumberValue(DistanceFromMoon);
        Values["speed"] = new NumberValue(Speed);
        Values["phase"] = new TextValue(Phase.Name);
    }

    public string FormattedValue(Metric metric, UnitSystem units)
    {
        return Values.TryGetValue(metric.Source, out var value) ? value.Format(metric.Format, units) : "\u2014";
    }

    public string FormattedDetailValue(Metric metric, UnitSystem units)
    {
        return Values.TryGetValue(metric.Source, out var value) ? value.FormatDetail(metric.Format, units) : "\u2014";
    }
}
